packages/wonder-blocks-layout/components/media-layout.js

```diff
@@ -52,104 +67,67 @@ type State = {|
-                delete watchHandlers[query];
+    componentDidMount() {
+        // TODO(WB-534): handle changes to mediaSpec prop
+        for (const [size, spec] of ((Object.entries(
+            this.props.mediaSpec,
+        ): any): Array<[MediaSize, {query: string}]>)) {
+            const mql = MEDIA_QUERY_LISTS[spec.query];
+            // during SSR there are no MediaQueryLists
+            if (!mql) {
+                continue;
             }
+            const listener = (e) => {
+                if (e.matches) {
+                    this.setState({size});
+                }
+            };
+            mql.addListener(listener);
+            this.cleanupThunks.push(() => mql.removeListener(listener));
         }
     }
 
-    getCurrentSize(spec: MediaSpec) {
+    componentWillUnmount() {
+        this.cleanupThunks.map((thunk) => thunk());
+    }
+
+    getCurrentSize(spec: MediaSpec): MediaSize {
         // If we have a state with the current size in it then we always want
         // to use that. This will happen if the viewport changes sizes after
         // we've already initialized.
         if (this.state.size) {
             return this.state.size;
-        }
-
-        // watchHandlers should never be undefined when state.size is also
-        // undefined, but just in case we fall back to the default size
-        if (this.watchHandlers) {
-            return DEFAULT_SIZE;
-        }
-
-        // We then go through and set up matchMedia watchers for each breakpoint
-        // (if they haven't been created already) and we add listeners to
-        // watch for when the viewport changes size.
-        this.watchHandlers = {};
-
-        let matchedSize;
-
-        for (const size of VALID_MEDIA_SIZES) {
-            // Flow has issues with checking !spec[size] directly so we store
-            // spec[size] in a new variable to get around the issue.
-            const _spec = spec[size];
-            if (!_spec) {
-                continue;
-            }
-
-            const {query} = _spec;
-
-            // Don't watch sizes that don't have an associated query
-            if (!query) {
-                continue;
-            }
-
-            // Create a new matchMedia watcher if one doesn't exist yet
-            // TODO(kevinb): move this to componentWillMount
-            if (!MediaLayout.WATCHERS[query]) {
-                MediaLayout.WATCHERS[query] = window.matchMedia(query);
-            }
-
-            const watcher = MediaLayout.WATCHERS[query];
-
-            // Attach a handler that watches for the change, saving a
-            // references to it so we can remove it later
-            const handler = (this.watchHandlers[query] = (e) => {
-                if (e.matches) {
-                    this.setState({size});
+        } else {
+            for (const [size, spec] of ((Object.entries(
+                this.props.mediaSpec,
+            ): any): Array<[MediaSize, {query: string}]>)) {
```

Comment:
Again, I'd like to see this split into a const and the loop to make the code and typing a little easier on the eye.